### ISSUE TITLE
VirtualBox install attempt causes [FAIL] FATAL: Detected Missing Required CPU Feature: Carry-less Multiplication (pclmulqdq)

### ISSUE THREAD

**Issue by danintel** [clearlinux/distribution#2161]
I get this error installing Clear Linux clear-33590-live-server.iso with VirtualBox 6.0 and 6.1:
`
[FAIL] FATAL: Detected Missing Required CPU Feature: Carry-less Multiplication extensions (pclmulqdq)`

My laptop supports PCLMULQDQ and other features (i7-8665U CPU), but VirtualBox does not. Unfortunately, for some reason, Clear Linux does not detect this "missing" feature at boot time and execute another code path (without the PCLMULQDQ). I realize PCLMULQDQ (Carryless multiply) optimizes the GCM part of AES-GCM, but maybe having an alternate code path for installs on VirtualBox VMs (and older computers and possibly other VMs) would make adoption of Clear Linux possible.

There is documentation on installing Clear Linux on VirtualBox, so Clear Linux install on VB seems to have worked sometime in the past (the document shows clear-29400-live-server being installed on 5/2019):
_Clear Linux* OS on VirtualBox*_
https://docs.01.org/clearlinux/latest/get-started/virtual-machine-install/virtualbox-cl-installer.html

**Comment by danintel** [comment]
Similar issue (but with bare metal apparently) with issue #1624

**Comment by thiagomacieira** [comment]
Clear Linux does not provide a fallback code path without PCLMULQDQ.

Can you tell exactly what's printing that message? Is it the installer?

**Comment by danintel** [comment]
@thiagomacieira Attached screenshot of error message while booting live DVD:

![VirtualBox_taylor_14_10_2020_17_18_3=fail-fatail-missing-pclmulqdq](https://user-images.githubusercontent.com/32893509/96058987-93302080-0e41-11eb-89d7-728c0f0c135d.png)

**Comment by thiagomacieira** [comment]
The reason for this is that Clear Linux default to optimising for Westmere (the shrink of the original Intel Core-i7/i5/i3 CPUs, from 2010) and for them, GCC thinks PCLMULQDQ is active by default.
```
$ gcc -dM -E -xc /dev/null | grep MUL
#define __PCLMUL__ 1
```
That means all throughout Clear, the compiler and libraries can generate use of this instruction without checking for runtime support.

Sorry, I don't think we can fix this. You need to get VirtualBox to enable the CPUID bit or use a different hypervisor.

**Comment by nuket** [comment]
I have the same exact issue on an Ivy Bridge i7 and VirtualBox 6.1.20.
But it does work within the Hyper-V that comes with Windows 20H2 (19042).

**Comment by thiagomacieira** [comment]
VirtualBox leaves too many CPUID features disabled. I discovered two weeks ago that it doesn't enable FMA either.

Sorry, you should really get a better hypervisor. Won't fix.

**Comment by nuket** [comment]
Lol. This is arrogant as f--k. Don't shoot the messenger.

I said I tried it with Hyper-V and it was fine, fastest Linux boot to desktop I've seen actually, but categorically saying that everyone using VirtualBox is not a worthy of investment is hilarious. I guess you think the only audience you want to serve is the datacenter? Perhaps Intel should invest in patching up VirtualBox to expose those features and gift people a zero-hassle way of getting things done.

Or talk to the people _at Intel_ who build [HAXM](https://github.com/intel/haxm) and have them build a zero-footprint live image runner. It's called feeding the pipeline of potential users, who also now have to learn your way of doing software package installs with swupd (which isn't apt or yum or rpm). Awesome.

I have no incentive to continue trying your fully-optimized version of Linux even if it does win benchmark shootouts left and right, and would love to discuss with whoever is running the open source division at Intel that you don't want to move the mountain to Moses.

**Comment by thiagomacieira** [comment]
The proper fix should be to fix VirtualBox, indeed. I've already raised this with our VM team, though there hasn't been progress yet.

Reducing the capability of Clear Linux because one hypervisor hasn't enabled a 12-year-old technology is not going to happen.

**Comment by Jonadabe** [comment]
Hum... the same here. :-(

**Comment by quiescens** [comment]
TLDR for the benefit of people that ended here searching for the same problem:
VBoxManage setextradata "vm name here" VBoxInternal/CPUM/IsaExts/PCLMUL 1

As far as both sides of this argument:

VirtualBox probably doesn't want to aggressively advertise newer host CPU features to guests because when it becomes necessary to move a VM to a CPU with a different featureset it can cause problems.
However, the interface (no GUI) and documentation of manually enabling something like this is very much lacking - I had to dig through the source code to find this one.
Even if VirtualBox doesn't want to add PCLMUL to the default Linux config, they could probably consider adding "Clear Linux" as one of version variants and set the flag automatically when that option is chosen.

Clear Linux doesn't want to hamstring their builds to support the lack of a 10 year old CPU flag, which is a reasonable objection.
In practice the PCLMUL instructions are supported (even without the above config change), Virtualbox is simply not advertising the support to the guest.
Its only the test that is failing (if you remove the flag after installing, it will continue to work) - admittedly that is Virtualbox's fault but Clear Linux could probably do better here too, possibly offer a boot option that overrides these checks or updating the "Clear Linux OS on VirtualBox" documentation to include the command above.
